Casbah 2.8.2 upgrade - leverage MongoException while supporting legacy error handling.

noahlz opened the pull request:
@sief please check this alternative to #168. I added support for error handling via `MongoException` while providing some semblance of support for old code that relies on the debugging added by `SalatDAOError`s as well as code that uses `MongoConnection` (which I found - being a deprecated class - creates connections that do _not_ throw `MongoException` at all - you have to check last error!).

noahlz committed:
Issue #158 (Casbah 2.8.2 upgrade) provide some semblance of suport for legacy code that still uses SalatDAOErrors and/or MongoConnection.

noahlz committed:
Issue #158 (Casbah 2.8.2 upgrade) comment out broken test - SalatDAOs created from MongoConnection apparently do not (?) return a WriteResult with any errors for duplicate key inserts.

sief commented:
@noahlz Interesting what you found out about `MongoConnection`, and good to have this legacy safety net. Considering this legacy problem it does indeed make sense to have the SalatDaoError. 
Just one little thing: can you replace `var wr: WriteResult = null` with a val inside the `try` block? In the `catch` block it is `null` anyway.

BTW: what about `Either[WriteResult, MongoException]` as alternative to the `MongoErrorAdapter`?  (or the scalaz disjunction \/ )

noahlz commented:
It turns out that when you use `MongoConnection` duplicate entries provide no error indicator at all and fail "silently" (wtf!), meaning that `MongoClient` is a must with this change.

I had to push another commit that skips the (failing) `DeprecatedAlphaDAO` test.

I have to wonder if I'm missing something here, and there is actually some way to force a "legacy" duplicate key error. I tried forcing an unique index with `collection.createIndex` - to no avail.

`Either` is not appropriate here, because even if a `MongoException` occurs we still have a `WriteResult` - I would like to have both available to the developer inspecting the `SalatDAOError` even when `MongoException` occurs.

sief commented:
Don't know what I'm missing here, but from what I see `wr` will always be `null` in case of a `MongoException`. 

This `MongoConnection` behaviour is really weird, maybe a weak WriteConcern?

noahlz commented:
I tried explicitly setting the WriteConcern to SAFE and still MongoConnection didn't yield any useful error. (Perhaps I am - or the existing tests are - doing something else wrong)

noahlz committed:
Issue #158 (Casbah 2.8.2 upgrade) added more detail to tests that fail spuriously

noahlz committed:
Issue #158 (Casbah 2.8.2 upgrade) implemented @sief's suggestion that we just use Either[WriteResult, MongoException].

Also, added "isolated" directive to SalatDAOSpec to solve spurious test failures.

noahlz commented:
Changed that silly MongoErrorAdapter to a `Either` like you suggested and changed `SalatDAOSpec` so it doesn't fail randomly - the tests have to be both sequential and isolated.

One outstanding todo is to test this change with Mongo 2.6 just to be sure.  I tested it with 3.2 (Docker is great for this, btw)

noahlz commented:
I did some more work on this. You can check the "err" field for the error, since `getCachedLastError` is now (apparently) useless:

MongoClient connections throw MongoException:

```
scala> AlphaDAO.insert(List(alpha.copy(id = 0), alpha.copy(id = 1)))
res0: List[Option[Int]] = List(Some(0), Some(1))

scala> AlphaDAO.insert(List(alpha.copy(id = 1), alpha.copy(id = 2)))
com.novus.salat.dao.SalatInsertError:

    SalatDAO[Alpha,int](alpha_dao_spec): insert failed!

    Collection: alpha_dao_spec
    WriteConcern: WriteConcern { "getlasterror" : 1} / (Continue on error? false)
    Error: com.mongodb.MongoException$DuplicateKey: { "serverUsed" : "127.0.0.1:27017" , "ok" : 1 , "n" : 0 , "err" : "insertDocument :: caused by :: 11000 E11000 duplicate key error index: test_salat.alpha_dao_spec.$_id_  dup key: { : 1 }" , "code" : 11000}


FAILED TO INSERT ONE OR MORE OF YOUR DBOs (SEE Error FOR DETAILS):
{ "_typeHint" : "com.novus.salat.test.dao.Alpha" , "_id" : 1 , "beta" : [ { "_typeHint" : "com.novus.salat.test.dao.Delta" , "y" : "delta7a" , "z" : "delta7b"}]}
{ "_typeHint" : "com.novus.salat.test.dao.Alpha" , "_id" : 2 , "beta" : [ { "_typeHint" : "com.novus.salat.test.dao.Delta" , "y" : "delta7a" , "z" : "delta7b"}]}


  at com.novus.salat.dao.SalatDAO.insert(SalatDAO.scala:382)
  ... 43 elided
Caused by: com.mongodb.MongoException$DuplicateKey: { "serverUsed" : "127.0.0.1:27017" , "ok" : 1 , "n" : 0 , "err" : "insertDocument :: caused by :: 11000 E11000 duplicate key error index: test_salat.alpha_dao_spec.$_id_  dup key: { : 1 }" , "code" : 11000}
  at com.mongodb.CommandResult.getWriteException(CommandResult.java:88)
  at com.mongodb.CommandResult.getException(CommandResult.java:79)
  at com.mongodb.DBCollectionImpl.translateBulkWriteException(DBCollectionImpl.java:316)
  at com.mongodb.DBCollectionImpl.insert(DBCollectionImpl.java:191)
  at com.mongodb.DBCollectionImpl.insert(DBCollectionImpl.java:167)
  at com.mongodb.DBCollection.insert(DBCollection.java:93)
  at com.mongodb.casbah.MongoCollectionBase$class.insert(MongoCollection.scala:654)
  at com.mongodb.casbah.MongoCollection.insert(MongoCollection.scala:1162)
  at com.novus.salat.dao.SalatDAO.insert(SalatDAO.scala:383)
  ... 43 more
```

With legacy safety net: MongoConnection connections return a WriteResult with an "err" field that we can access via `getError`

```
scala> DeprecatedAlphaDAO.insert(List(alpha.copy(id = 1), alpha.copy(id = 2)))
com.novus.salat.dao.SalatInsertError:

    SalatDAO[Alpha,int](alpha_dao_spec): insert failed!

    Collection: alpha_dao_spec
    WriteConcern: WriteConcern { "getlasterror" : 1} / (Continue on error? false)
    Error: { "serverUsed" : "127.0.0.1:27017" , "connectionId" : 1098 , "err" : "insertDocument :: caused by :: 11000 E11000 duplicate key error index: test_salat.alpha_dao_spec.$_id_  dup key: { : 1 }" , "code" : 11000 , "n" : 0 , "ok" : 1.0}


FAILED TO INSERT ONE OR MORE OF YOUR DBOs (SEE Error FOR DETAILS):
{ "_typeHint" : "com.novus.salat.test.dao.Alpha" , "_id" : 1 , "beta" : [ { "_typeHint" : "com.novus.salat.test.dao.Delta" , "y" : "delta7a" , "z" : "delta7b"}]}
{ "_typeHint" : "com.novus.salat.test.dao.Alpha" , "_id" : 2 , "beta" : [ { "_typeHint" : "com.novus.salat.test.dao.Delta" , "y" : "delta7a" , "z" : "delta7b"}]}


  at com.novus.salat.dao.SalatInsertError$.apply(SalatDAOErrors.scala:80)
  at com.novus.salat.dao.SalatDAO$$anonfun$insert$3.apply(SalatDAO.scala:385)
  at com.novus.salat.dao.SalatDAO$$anonfun$insert$3.apply(SalatDAO.scala:385)
  at com.novus.salat.dao.SalatDAO.handleLegacyErrors(SalatDAO.scala:350)
  at com.novus.salat.dao.SalatDAO.insert(SalatDAO.scala:386)
  ... 43 elided
```

Because Mongo stops processing the list of DBOs at the first error (assuming you have halt on error enabled), it still might have inserted some records before hitting the error.  So, I made the exception text slightly more ambiguous on what DBOs failed when there are more than one (bulk insert).

noahlz committed:
Issue #158 (Casbah 2.8.2 upgrade) handle legacy errors when user is using MongoConnection. With unit tests.

Also, reworded some of the SalatDAOError messages to be more ambiguous about DBO operations (insert, update) "failing." Mongo might actually insert some DBOs but not others (you have to check the error message)

noahlz committed:
Issue #158 reword new unit tests.

noahlz commented:
```
[info] Salat simple DAO should
[info] + supply a useful description for debugging
[info] + insert a case class
[info] + insert a collection of case classes
[info] + no-op inserting an empty collection of objects
[info] + handle MongoExceptions when inserting an object having a duplicate key
[info] + handle MongoExceptions for bulk inserts where an object has a duplicate key
[info] + handle legacy errors when inserting an object having a duplicate key
[info] + handle legacy errors for bulk inserts where an object has a duplicate key
[info] + handle MongoExceptions for invalid updates
[info] + handle legacy errors for invalid updates
```

Ok, I'm going to leave this change here for a few days for review, and then merge into the `1.10.x-branch` unless there's strong objection. Possibly cut `1.10.0` final next weekend.

noahlz committed:
Issue #158 touch up on naming and comments.

noahlz committed:
Issue #158 missed a spot.
